python/templates/macros/implementations.jinja2

```diff
@@ -99,6 +100,11 @@ void {{ class_type }}::{{ relation.setter_name(get_syntax) }}({{ relation.full_t
 {% for relation in relations %}
 {% if with_adder %}
 void {{ class_type }}::{{ relation.setter_name(get_syntax, is_relation=True) }}({{ relation.full_type }} component) {
+  if (m_obj->data.{{ relation.name }}_end != m_obj->m_{{ relation.name }}->size()) {
```

Comment:
Can you add a comment here that says why we are doing this for the future? Should we just make this copy in the `clone` call directly without waiting for it to trigger here? The effect would be the same, but I think the code would be easier to follow.

Reply:
I think in this case I would go for the implementation in `clone` because it's easier to see why it happens there. We can always come back and move it to a slightly more unintuitive place once we see performance is actually an issue. In the end for most relations we are copying a vector of pointers and most vector members are also pretty small objects.

Reply:
I had a look at saving cloned objects and that's worse when it's implemented in the `addX` because after cloning each element will point to the big vector and for each element saved a big vector will also be saved (even though when reading since we have the indexes it's fine). (For example, if element 0 has a relation to 'a', 1 to 'b', 2 to 'c' and 3 to 'd', if 0 is cloned and saved then ['a', 'b', 'c', 'd'] is saved but if 0 and 1 are both cloned then ['a', 'b', 'c', 'd', 'a', 'b', 'c', 'd'] is saved. So certainly having it in `addX` with the same behaviour as in `clone()` is more complicated

Reply:
Yeah, I had this though over lunch as well. I think `clone` would have to copy contents to a different vector in order to make the writing side work again as well. I am currently having a look to remember which one should actually be used for this.

Reply:
OK. I think the additional vector I was thinking about doesn't really play a role here, it's in `CollectionData` and should be handled correctly if we make sure to handle the vector correctly here.

In that case I agree with you and adding things in `clone` is the easier way, because otherwise we have to keep track of whether `addX` has been called and then do the copy in any case before writing in order for the rest to work as expected.

Reply:
> In that case I agree with you and adding things in `clone` is the easier way, because otherwise we have to keep track of whether `addX` has been called and then do the copy in any case before writing in order for the rest to work as expected.

I think no tracking needs to be done here. The workflow would be to have this fix in `addX`, then when writing we can have objects that are pointing to a big vector but we may not want to write it all and objects that are pointing to a "small" vector so we have to use the `X_begin` and `X_end` indexes to create a new big vector to only write what the objects are pointing to. Probably starting here: https://github.com/AIDASoft/podio/blob/master/python/templates/CollectionData.cc.jinja2#L134